File: no-consecutive-ones-600.cpp
class Solution {
public:
    int findIntegers(int n) {
        int fib_len = log2(n) + 2;
        int fibs[fib_len];
        fibs[0] = 1;
        fibs[1] = 2;
        int a = 1;
        int b = 2;
        for (int i = 2; i < fib_len; i++) {
            int c = a + b;
            a = b;
            b = c;
            fibs[i] = b;
        }
        int count = 0;
        while (n >= 0) {
            if (n <= 1) {
                count += n + 1;
                break;
            }
            if (n == 3) {
                count += 3;
                break;
            }
            int mag = (int)log2(n);
            count += fibs[mag];
            if (n & (int)pow(2, mag - 1)) {
                count += fibs[mag - 1];
                break;
            }
            else {
                n -= pow(2, mag);
            }
        }
        return count;
    }
};
//0 ms (100%ile), 5.9 MB (69%ile)

File: no-consecutive-ones-600.py
from math import log2

class Solution:
    def findIntegers(self, n):
        fibs = [1, 2]
        append = fibs.append
        a, b = 1, 2
        for i in range(int(log2(n))):
            a, b = b, a+b
            append(b)
        count = 0
        while n >= 0:
            if n <= 1:
                count += n + 1
                break
            if n == 3:
                count += 3
                break
            mag = int(log2(n))
            count += fibs[mag]
            if n & (2 ** (mag - 1)):
                count += fibs[mag - 1]
                break
            else:
                n -= 2 ** mag
        return count
#32 ms (90%ile), 14.2 MB (61%ile)
